File: cpp/horner.cpp
/*
 * horner.cpp
 */


#include <iostream>

using namespace std;

// W(x) = 2x^3 + 3x^2 + 5x + 4
// W(x) = x(x^2 + 3x + 5) + 4
// W(x) = x(x(x+3) + 5) + 4

float horner_it(int n, float tab[], float x)
{
    float wynik = tab[0];
    for (int i=1; i < n+1; i++)
    {
        wynik = wynik* x + tab[i];
    }
    return wynik;
}




int main(int argc, char **argv)
{
    float x;
    int n;
    float tab[n];
    
    cout<< "Podaj argument." <<endl;
    cout<< "x=";
    cin>> x;
    cout<< "Podaj stopień wielomianu."<<endl;
    cin>> n; 
    for (int i = 0; i < n; i++)
    {
        cout<< "Podaj współczynniki."<< endl;
        cin>> tab[i];
    }

    cout<< "Wartość wielomianu: " << horner_it(n, tab, x) <<endl;
    
	return 0;
}

File: python/horner.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
#  horner.py


def horner_it(stopien, tb, x):
    wynik = tb[0]
    for i in range(1, stopien + 1):
        wynik = wynik * x + tb[i]

    return wynik


def horner_rek(a, tb, x):
        # a- stopień, tb- tablica
    if a == 0:
        return tb[0] * x
    return horner_rek(a - 1, tb, x) * x + tb[a]


def main(args):
    tb = []
    stopien = 3
    x = int(input("Podaj wartośc x: "))
    for i in range(0, 4):
        tmp = int(input("Podaj współczynniki: "))
        tb.append(tmp)

    print("Wynik: ", horner_rek(stopien, tb, x))
    return 0


if __name__ == '__main__':
    import sys
    sys.exit(main(sys.argv))
